### ISSUE TITLE
distcc tcp mode is a security risk [CVE 2004-2687]

### ISSUE THREAD

**Issue by sourcefrog** [distcc/distcc#155]
http://www.cvedetails.com/cve/2004-2687

distcc currently has two modes by which clients can connect - over TCP (default) or SSH. 

In TCP mode, distcc checks the client IP address against a whitelist, which (iirc) is required but can be set quite loosely. There is of course no guarantee that every user on a permitted client address is friendly.

Once the connection is established the client can reasonably easily manipulate the server into running arbitrary commands. This is automated in eg http://www.rapid7.com/db/modules/exploit/unix/misc/distcc_exec

Possible solutions include:
- Deprecating TCP mode and requiring SSH connections. This would largely push the problem onto SSH and make it more reasonable for distccd to trust the client.  However SSH does have a noticeable performance overhead, or to be more precise it did when I measured 10 years ago. It may have changed. It may cause problems with server discovery.
- Use some other transport that provides authentication/integrity/encryption but is faster than SSH.
- Add a simple authentication protocol. This may be cheaper than SSH. There is a risk the authentication protocol itself would be buggy. If there's only authentication of thie client without integrity checks or encryption, it may remain vulnerable to malicious networks.
- Using platform-specific security features such as `seccomp_bpf` to restrict what can be done once the compiler command is launched. This might be a useful adjunct to other measures but is probably not enough, and may be brittle.

**Comment by Mathiasdm** [comment]
I would argue that, within a corporate network, it may be preferable to go with the TCP mode (trusting the local machines and users) to gain the best performance.

**Comment by stevenhoneyman** [comment]
Or private isolated (i.e. trusted) network. I have no need for any security in distcc - if someone manages to get close enough to touch that, they already have complete physical access anyway!

Optional extra, great. Mandatory... not so much.

**Comment by onionjake** [comment]
You can also use SSH multiplexing to only pay the overhead of establishing the ssh connection once: http://www.linuxjournal.com/content/speed-multiple-ssh-connections-same-server

The server will limit how many simultaneous sessions per network connection, so keep that in mind: http://unix.stackexchange.com/a/22987

**Comment by uzytkownik** [comment]
> Add a simple authentication protocol. This may be cheaper than SSH. There is a risk the authentication protocol itself would be buggy. If there's only authentication of thie client without integrity checks or encryption, it may remain vulnerable to malicious networks.

Or just use TLS + SASL. That is essentially solved problem and a lot of effort is done to make it both fast and secure for browsers.

> Using platform-specific security features such as seccomp_bpf to restrict what can be done once the compiler command is launched. This might be a useful adjunct to other measures but is probably not enough, and may be brittle.

That secures only one side of connection. Other problem is 'trusting trust' like attacks - the MiTM pretends to be server but sends a malicious result. 

> Or private isolated (i.e. trusted) network. I have no need for any security in distcc - if someone manages to get close enough to touch that, they already have complete physical access anyway!

This assumes no 'road-warrior' type of setup. If I have a laptop I don't want to switch config every time I leave safe network as I do it few times a day. If I don't every time I'm in coffee shop and try to compile anyone can exploit that. Even worse the trusted network can be broken without physical access if, for example, WPA2 was broken - [which is no longer hypotetical](https://en.wikipedia.org/wiki/KRACK).

**Comment by shawnl** [comment]
> Using platform-specific security features such as seccomp_bpf to restrict what can be done once the compiler command is launched.

I added a seccomp filter https://github.com/distcc/distcc/pull/235 , but the actual binary that gets loaded (ARGV[0]) can still be set by the user. There needs to be a whitelist. I could bring Debian's update-distcc-symlinks script in-house (and add clang support), and then make that the whitelist.

**Comment by sourcefrog** [comment]
> Or just use TLS + SASL. That is essentially solved problem and a lot of effort is done to make it both fast and secure for browsers.

Reusing TLS + SASL is great, but it does not just magically solve the problem. The typical deployment of TLS relies on a CA root to authenticate the server. People probably won't get an authenticated certificate for their internal build farms. TLS outside of browsers has very commonly been done with no meaningful validation of the server, and as you note that has real problems. (see eg TLS in SMTP.) Let's not reproduce them here.

**Comment by uzytkownik** [comment]
>> Or just use TLS + SASL. That is essentially solved problem and a lot of effort is done to make it both fast and secure for browsers.
> Reusing TLS + SASL is great, but it does not just magically solve the problem. The typical deployment of TLS relies on a CA root to authenticate the server. People probably won't get an authenticated certificate for their internal build farms.

I don't think this is necessarily true. You can have self-signed certificates that are white-list or just create a single CA cert for farm and trust it - no need to get an authenticated certificate (which in age of letsencrypt is not that hard anyway).

> TLS outside of browsers has very commonly been done with no meaningful validation of the server, and as you note that has real problems. (see eg TLS in SMTP.) Let's not reproduce them here.

Err. I fail to see how this is property of protocol although yes, you need to ensure that you use the APIs correctly - _though it is still probably easier to [get API right than crypto](https://www.nccgroup.trust/us/about-us/newsroom-and-events/blog/2009/july/if-youre-typing-the-letters-a-e-s-into-your-code-youre-doing-it-wrong/)_.

That said - I have absolutely no idea what real problems you are referring to. SMTP/POP3/IMAP has TLS support since 'ever' and usually it is the only way to access those services (port 25 is routinely blocked by ISPs) as is Jabber, IRC, NNTP, LDAP and many other services - and yes, they do check the certificates. It's true that MTA -> MTA SMTP is problematic but this has to do more with historical reasons (blank enforcement would break communication overnight and adoption is slow) than TLS per se. Could you elaborate what problems you are referring to?

**Comment by shawnl** [comment]
@sourcefrog Can you review my whitelist approach to the problem? It is actual code, rather than speculation.

**Comment by sourcefrog** [comment]
@shawnl done, thanks for the PR!

@uzytkownik We just need to think about the overall experience for the user and whether it is likely to give practical security. For example the distcc docs would tell them to

 - install openssl
 - make a CA
 - use that CA to make a CSR per helper machine, matching the IP or name through which it's addressed
 - sign those CSRs
 - put the signed certificate onto each helper
 - configure the CA to be trusted by each client, and that the client must see a valid cert

Or perhaps they use letsencrypt instead (a great project) - but doesn't that require the server be reachable? It seems to be a problem if the helper machines are either not reachable on the internet or even more so if they have only rfc1918 addresses, and both may be good ideas.
 
oh but wait also
 - decide what sasl mechanism to use
 - generate probably also per-client keypairs
 - put the relevant keys onto all clients and helpers

The process is complicated enough I'm afraid many people either won't bother (thinking they have network-level protection) or won't get it right, and so the overall security state won't move enough.

This isn't an argument for rolling our own crypto, or an argument about the cryptographic or protocol-design merits of TLS.

TLS in SMTP being possible in the protocol, but rarely actually enforced or checked is the perfect example of what we should avoid. The other TLS fiasco I was thinking of is https://www.python.org/dev/peps/pep-0476/.

**Comment by shawnl** [comment]
@sourcefrog actually it is here: https://github.com/distcc/distcc/pull/243

**Comment by shawnl** [comment]
@onionjake
you can also use ssh -D

**Comment by uzytkownik** [comment]
@sourcefrog 

> We just need to think about the overall experience for the user and whether it is likely to give practical security.

Yeah. Security is hard. Nonetheless I view those issues as orthogonal to protocol - for example you don't _need_ CA if you use self-signed certificates (i.e. certificate chain of size 1). You can imagine that distcc would generate a certificate on first use via distribution scripts or something. On opposite side distcc might [trust certificate on first use (as ssh does)](https://en.wikipedia.org/wiki/Trust_on_first_use). This has it's own set of problems but security is about tradeoffs. 

This is circumventing a part of TLS security, however, and I don't think it is unreasonable to make admins create CA and certs etc. if they are concerned about security as they are likely have some 'laying around' for other services (email, http, ldap...). If not Google shows ~1,000,000 results of how to generate self-signed cert using openssl so I don't think it needs to be documented by distcc (I'm quite sure I was able to generate it back when I was in High School).

> TLS in SMTP being possible in the protocol, but rarely actually enforced or checked is the perfect example of what we should avoid. The other TLS fiasco I was thinking of is https://www.python.org/dev/peps/pep-0476/.

TLS is not a magic bullet if you don't enforce it ;). However MUA to MTA TLS was enforced by any server I used for about decade. MTA to MDA SMTP transport was historically considered secure as the internet backbone was consider secure. Without getting too political here some companies started fearing state actors so they started introducing encryption but adoption was slow. However this is different with distcc as distcc servers don't form a one big network as SMTP servers do.

Those are more issues of implementation (and in latter case wrong API defaults) than TLS per se. In similar way you would need to decide on security level as in case of SMTP. To replace TLS by something you would need to have correct protocol AND implementation, which is harder. I don't think there is some ready as TLS is standard.

> This isn't an argument for rolling our own crypto

Note that 'rolling our own crypto' only about rolling own cryptography primitives but also protocol. Even though TLS is probably close to state-of-the-art [there](https://en.wikipedia.org/wiki/FREAK) [were](https://en.wikipedia.org/wiki/DROWN_attack) [multiple](https://en.wikipedia.org/wiki/Transport_Layer_Security#BEAST_attack) [problems](https://en.wikipedia.org/wiki/POODLE). 

> decide what sasl mechanism to use

I think this is up to the configuration, not something which is decided by app developer. I know that one some of my machines I'd like to use GSSAPI and on some CRAM-MD5. Trying to roll own auth is painful for admins as they need to get it synced with existing infrastructure somehow. Cyrus SASL has even standard config which decides for app. I'm not sure about other libraries though.

**Comment by shawnl** [comment]
Fixed with d849ca2bcf6

**Comment by sourcefrog** [comment]
It's not a complete fix, because a malicious client can still cause trouble
through options to the compiler.

**Comment by shawnl** [comment]
> a malicious client can still cause trouble through options to the compiler.

That is a much more difficult attack however, and would probably count as a bug in gcc or clang.

**Comment by sourcefrog** [comment]
>> a malicious client can still cause trouble through options to the compiler.

> That is a much more difficult attack however, and would probably count as a bug in gcc or clang.

> This means it requires 2 bugs (a local gcc or clang arbitrary code execution bug and a local privilege escalation bug) rather than just a privilege escalation bug.

Three things: (1) it doesn't need RCE in the compiler, and even if it did, (2) there's no reason to think RCEs things don't exist, and (3) giving someone access to the distcc daemon is a distcc security hole even if they then likely want a privilege escalation bug. (From the video https://youtu.be/bSkpMdDe4g4 you linked and the earlier commits I think you see this but I'll just be explicit.)

(1) If you think about what attackers typically want to do on the host under attack, it's to run a program, read a file, or write a file. All of those are available through gcc options or other means, without needing any compiler bugs. It's great that you block some in a316366 but this is always going to be whack-a-mole.

(2) Compilers run extremely complex code usually written in unsafe languages on complex input. They're not normally considered a trust boundary and I doubt they have had much security scrutiny. An RCE in gcc might be considered a bug there, but we should expect there are many such bugs.

Empirically, all the [CVEs in gcc](https://www.cvedetails.com/vulnerability-list/vendor_id-72/product_id-960/GNU-GCC.html) are issues in the code it generates, not in the compiler itself, which I take to mean people aren't even looking seriously for things that subvert the compiler.

You also have to remember that many distcc deployments are intentionally running quite old compilers, to target old OSes, and others will be just accidentally running old compilers because, again, it's not a trust boundary.

So, overall: filtering compiler options is great, but it was a huge mistake on my part not to authenticate the client, and it's one I would really urge should be fixed.

**Comment by TafThorne** [comment]
"distcc should authenticate clients to increase security" is not the same as "distcc should drop support for TCP to increase security".  It sounds like there are good reasons to offer both of those features as options, possibly even the options configured by default in some packages.  Should we close this lengthily ticket by creating several specific feature requests for the various options people would like to see one day?

**Comment by afbjorklund** [comment]
Looks like users have a hard time figuring out how to use the `/usr/sbin/update-symlinks` script,
even after installing `python3` to actually make it work... And the `/usr/lib/distcc` is hardcoded.

**Comment by gearhead** [comment]
Ok, I updated arch and this gets pushed. Try to compile a program on the headless raspberry pi and start distccd on the main computer. I get all the errors saying 'failed to distribute'. Stumbled on this issue and have reverted to the previous version of distcc on the main computer. It is running on my nat-ted LAN, I have the one IP address running distccd and one address listed as authorized. I only fire it up when I need to compile something. I understand and accept the risks. How do I enable the previous behavior?

**Comment by shawnl** [comment]
--make-me-a-botnet

**Comment by gearhead** [comment]
How does one allow a tcp connection with the current version of distcc? 
The current version on Arch is distcc 3.3-2. This version (with my previous configuration files) will not allow a tcp connection to the slave for compilation. If I downgrade it to 3.2-rc-1, everything compiles. If I upgrade to 3.3-2, it fails with authentication errors and the master gets connection failures. I get the security risk, but there should be a way of enabling the previous behavior. I have not found a way of doing that.

**Comment by shawnl** [comment]
Pass --make-me-a-botnet to distccd (you will have to edit the service file)

**Comment by shawnl** [comment]
I think the proper solution to this bug is to run distcc over wireguard https://www.wireguard.com/

**Comment by shawnl** [comment]
SSH also doesn't have to be so slow. If ssh is just used to forward to tcp mode distcc using -L (and -o "ServerAliveInterval 60" -o "ServerAliveCountMax 0" for a 60 second timeout so distcc's mostly-stateless nature can be preserved) it will be MUCH more performant (although not as performant as wireguard which is UDP---ssh is TCP).

**Comment by TafThorne** [comment]
> --make-me-a-botnet

That is not a very obvious way to allow TCP 

`--allow-tcp` or `--enable-tcp` or `--tcp-enable` would all be more obvious.  Security is an issue but when you have a private compile far, simplicity and performance are the main objectives.  If you wanted any or all of those to issue a warning when you run it of "must use the --make-me-a-botnet parameter too" that would be fine but right now the command does not tell users what it does.

**Comment by gearhead** [comment]
Agreed. I did not even understand that what he posted was actually a
configuration flag. I interpreted the response as a joke. Humor can work in
the commented source code, but rarely works on the command line or
configuration files...

Even a flag like --enable-tcp-unsecure may be more clear...

**Comment by ollieparanoid** [comment]
> If there are no objections I will implement --enable-tcp-insecure this evening after work.

+1 for that, since now this flag is used by the distcc client automatically when executing distccd via SSH. Having `--enable-tcp-insecure` scrolling by in htop by a childprocess of sshd sounds a lot more reasonable than `--make-me-a-botnet`.
